File: packages/engine/src/events.ts
import { EventIds } from './constants';

/**
 * Named tick durations for animation/timing of resolved event groups.
 * The runtime value is a plain `number` so groups can sum durations safely.
 */
export const Duration = {
	ZERO: 0,
	MAFIA_KILL: 3,
	SHOOTOUT: 3,
} as const;

export type GameEventTargets = string[];

export type GameEventDump = {
	eventId: string;
	targets: GameEventTargets;
	message: string;
};

export type GameEventGroupDump = {
	groupId: string | null;
	duration: number;
	events: Array<GameEventDump | GameEventGroupDump>;
};

export class GameEvent {
	constructor(
		public eventId: string,
		public targets: GameEventTargets,
		public message: string,
	) {}

	dump(): GameEventDump {
		return {
			eventId: this.eventId,
			targets: [...this.targets],
			message: this.message,
		};
	}
}

export type GameEventEntry = GameEvent | GameEventGroup;

export class GameEventGroup {
	public duration: number = Duration.ZERO;
	public events: GameEventEntry[] = [];

	constructor(public groupId: string | null = null) {}

	newEvent(event: GameEvent) {
		this.events.push(event);
	}

	newEventGroup(eventGroup: GameEventGroup) {
		this.duration += eventGroup.duration;
		this.events.push(eventGroup);
	}

	reset(newId?: string) {
		this.events = [];
		this.duration = Duration.ZERO;
		if (newId !== undefined) this.groupId = newId;
		return this;
	}

	clone(): GameEventGroup {
		const group = new GameEventGroup(this.groupId);
		group.duration = this.duration;
		group.events = this.events.map((event) =>
			event instanceof GameEventGroup
				? event.clone()
				: new GameEvent(event.eventId, [...event.targets], event.message),
		);
		return group;
	}

	dump(): GameEventGroupDump {
		return {
			groupId: this.groupId,
			duration: this.duration,
			events: this.events.map((event) => event.dump()),
		};
	}
}

/**
 * @deprecated Use {@link EventIds} from `@mafia/engine` directly.
 * Retained for backward compatibility with existing consumers.
 */
export const CommonEvents = {
	NIGHT_IMMUNE: EventIds.NIGHT_IMMUNE,
	KILLED_BY_MAFIA: EventIds.KILLED_BY_MAFIA,
} as const;

File: old/packages/engine/src/engine/events.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import List, Union


@dataclass
class Duration:
    ZERO: int = 0
    # Mafia actions
    MAFIA_KILL: int = 3
    # Town Actions
    SHOOTOUT: int = 3
    # Neutral Actions


@dataclass
class GameEvent:
    """What event was it, and who should it be broadcast to"""

    event_id: str
    targets: list
    message: str

    def dump(self) -> dict:
        return asdict(self)


@dataclass
class GameEventGroup:
    """A Grouping of game events, eg. Broadcast event A to all players, and event B to select players"""

    group_id: str = None
    duration: Duration = Duration.ZERO
    events: List[Union[GameEvent, GameEventGroup]] = field(default_factory=list)

    def new_event(self, event: GameEvent):
        self.events.append(event)
        return

    def new_event_group(self, event_group: GameEventGroup):
        self.duration += event_group.duration
        self.events.append(event_group)
        return

    def reset(self, new_id: str = None):
        self.events.clear()
        self.duration = 0
        if new_id:
            self.group_id = new_id
        return self

    def get_by_id(self, id):
        for event in self.events:
            if isinstance(event, GameEventGroup) and event.group_id == id:
                return event
            elif isinstance(event, GameEvent) and event.event_id == id:
                return event

    def dump(self) -> dict:
        return asdict(self)["events"]


# Create a root event group. Bit silly but I want to use the methods
# EVENTS = GameEventGroup(group_id='root')
ACTION_EVENTS = GameEventGroup(group_id="action")


# ------- Shared Events ------- #
@dataclass
class Common:
    INVALID_TARGET = "invalid_target"
    NIGHT_IMMUNE = "night_immune"
    KILLED_BY_MAFIA = "killed_by_mafia"
    VISITED_BY = "visited_by"
